experimental/metal/direct_command_buffer.m

```diff
@@ -337,50 +338,50 @@ iree_status_t iree_hal_metal_direct_command_buffer_create(
-        &iree_hal_metal_command_buffer_vtable, &command_buffer->base);
-    command_buffer->queue = [queue retain];  // +1
-    command_buffer->builtin_executable = builtin_executable;
-    iree_arena_initialize(block_pool, &command_buffer->arena);
-    command_buffer->staging_buffer = staging_buffer;
-    command_buffer->host_allocator = host_allocator;
-    status = iree_hal_resource_set_allocate(block_pool, &command_buffer->resource_set);
-    iree_hal_metal_command_segment_list_reset(&command_buffer->segments);
-    @autoreleasepool {  // Use @autoreleasepool to trigger the autorelease within encoder creation.
-      // We track resource lifetime by ourselves in IREE; so just do unretained references to
-      // resources in Metal command buffer, which avoids overhead and gives better performance.
-      MTLCommandBufferDescriptor* descriptor = [MTLCommandBufferDescriptor new];  // +1
-      descriptor.retainedReferences =
-          resource_reference_mode == IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_RETAINED;
-      descriptor.errorOptions = MTLCommandBufferErrorOptionNone;
-      command_buffer->command_buffer =
-          [[queue commandBufferWithDescriptor:descriptor] retain];  // +1
-      [descriptor release];                                         // -1
-    }
-    const iree_hal_metal_device_params_t* params = iree_hal_metal_device_params(device);
-    command_buffer->dispatch_type =
-        params->command_dispatch_type == IREE_HAL_METAL_COMMAND_DISPATCH_TYPE_CONCURRENT
-            ? MTLDispatchTypeConcurrent
-            : MTLDispatchTypeSerial;
-    command_buffer->state.compute_encoder = nil;
-    command_buffer->state.blit_encoder = nil;
-    command_buffer->state.encoder_event = [queue.device newEvent];  // +1
-    command_buffer->state.next_encoder_event_value = 1;
-    memset(command_buffer->state.descriptor_sets, 0, sizeof(command_buffer->state.descriptor_sets));
-    memset(command_buffer->state.push_constants, 0, sizeof(command_buffer->state.push_constants));
-
-    *out_command_buffer = &command_buffer->base;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_allocator_malloc(host_allocator, sizeof(*command_buffer), (void**)&command_buffer));
+
+  iree_hal_command_buffer_initialize(device, mode, command_categories, IREE_HAL_QUEUE_AFFINITY_ANY,
+                                     binding_capacity, &iree_hal_metal_command_buffer_vtable,
+                                     &command_buffer->base);
+  command_buffer->queue = [queue retain];  // +1
+  command_buffer->builtin_executable = builtin_executable;
+  iree_arena_initialize(block_pool, &command_buffer->arena);
+  command_buffer->staging_buffer = staging_buffer;
+  command_buffer->host_allocator = host_allocator;
+  iree_status_t status = iree_hal_resource_set_allocate(block_pool, &command_buffer->resource_set);
+  iree_hal_metal_command_segment_list_reset(&command_buffer->segments);
```

Comment:
you'll still want to guard all this below in is_ok(status) so that you don't set out_command_buffer/ref the staging buffer and instead release the command buffer before returning

Reply:
Ah, missed this part when refactoring.